ui/raidboss/emulator/data/LogEventHandler.ts

```diff
@@ -0,0 +1,119 @@
+import LineEvent from './network_log_converter/LineEvent';
+import { LineEvent0x01 } from './network_log_converter/LineEvent0x01';
+
+export default class LogEventHandler extends EventBus {
+  static doesLineMatch(line: string, regexes: Record<string, RegExp>): RegExpExecArray | false {
+    for (const i in regexes) {
+      const res = regexes[i]?.exec(line);
+      if (res) {
+        if (LocaleNetRegex.areaSeal[i as Lang] && res.groups)
+          res.groups.language = i;
+        return res;
+      }
+    }
+    return false;
+  }
+
+  static isMatchStart(line: string): RegExpMatchArray | false {
+    let res;
+    res = LogEventHandler.doesLineMatch(line, EmulatorCommon.countdownRegexes);
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.StartIn = (parseInt(res.groups.time ?? '0') * 1000).toString();
+      res.groups.StartType = 'Countdown';
+      return res;
+    }
+    res = LogEventHandler.doesLineMatch(line, EmulatorCommon.sealRegexes);
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.StartIn = '0';
+      res.groups.StartType = 'Seal';
+      return res;
+    }
+    res = LogEventHandler.doesLineMatch(line, EmulatorCommon.engageRegexes);
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.StartIn = '0';
+      res.groups.StartType = 'Engage';
+      return res;
+    }
+    return false;
+  }
+
+  static isMatchEnd(line: string): RegExpMatchArray | false {
+    let res;
+    res = LogEventHandler.doesLineMatch(line, { '': EmulatorCommon.winRegex });
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.EndType = 'Win';
+      return res;
+    }
+    res = LogEventHandler.doesLineMatch(line, { '': EmulatorCommon.wipeRegex });
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.EndType = 'Wipe';
+      return res;
+    }
+    res = LogEventHandler.doesLineMatch(line, { '': EmulatorCommon.cactbotWipeRegex });
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.EndType = 'Cactbot Wipe';
+      return res;
+    }
+    res = LogEventHandler.doesLineMatch(line, EmulatorCommon.unsealRegexes);
+    if (res) {
+      res.groups = res.groups ?? {};
+      res.groups.EndType = 'Unseal';
+      return res;
+    }
+    return false;
+  }
+
+  public currentFight: LineEvent[] = [];
+  public currentZoneName?: string;
+  public currentZoneId = -1;
+
+  parseLogs(logs: LineEvent[]): void {
+    for (let i = 0; i < logs.length; ++i) {
+      const lineObj = logs[i] as LineEvent;
+
+      this.currentFight.push(lineObj);
+
+      lineObj.offset = lineObj.timestamp - (this.currentFight[0]?.timestamp ?? 0);
+
+      const res = LogEventHandler.isMatchEnd(lineObj.networkLine);
+      if (res) {
+        this.endFight();
+      } else if (lineObj instanceof LineEvent0x01) {
+        const castLineObj = lineObj;
+        this.currentZoneId = parseInt(castLineObj.zoneId ?? '0');
+        this.currentZoneName = castLineObj.zoneName;
```

Comment:
```suggestion
        this.currentZoneId = parseInt(lineObj.zoneId ?? '0');
        this.currentZoneName = lineObj.zoneName;
```

Reply:
I could have sworn I tried this at one point and tsc complained, but it seems fine now? 🤷‍♂️

Adjusted for next commit.